quantization/BaseProductQuantization.h

```diff
@@ -0,0 +1,502 @@
+#pragma once
+
+#include <algorithm>
+#include <cassert>
+#include <cereal/access.hpp>
+#include <cereal/cereal.hpp>
+#include <cereal/types/memory.hpp>
+#include <cereal/types/vector.hpp>
+#include <cmath>
+#include <cstdint>
+#include <cstring>
+#include <flatnav/DistanceInterface.h>
+#include <memory>
+#include <quantization/CentroidsGenerator.h>
+#include <quantization/Utils.h>
+#include <queue>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace flatnav::quantization {
+
+using flatnav::quantization::CentroidsGenerator;
+
+template <typename n_bits_t> struct PQCodeManager {
```

Comment:
Why do we need this? I don't really understand what this does.

Reply:
Technically, we don't really need this. I added it jus to separate the logic that computes quantization codes from the general PQ logic, but you're right, it might be a bit confusing. I'll clean it up.